runtime/serde/serde-json/common/src/aws/smithy/kotlin/runtime/serde/json/JsonLexer.kt

```diff
@@ -0,0 +1,433 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package aws.smithy.kotlin.runtime.serde.json
+
+import aws.smithy.kotlin.runtime.serde.*
+import aws.smithy.kotlin.runtime.util.*
+
+private val DIGITS = ('0'..'9').toSet()
+private val EXP = setOf('e', 'E')
+private val PLUS_MINUS = setOf('-', '+')
+
+private typealias StateStack = ListStack<LexerState>
+private typealias StateMutation = (StateStack) -> Unit
+
+/**
+ * Manages internal lexer state
+ *
+ * The entire lexer works off peeking tokens. Only when nextToken() is called should state be mutated.
+ * State manager helps enforce this invariant.
+ */
+private data class StateManager(
+    private val state: StateStack = mutableListOf(LexerState.Initial),
+    private val pendingMutations: MutableList<StateMutation> = mutableListOf()
+) {
+
+    /**
+     * The size of the state stack
+     */
+    val size: Int
+        get() = state.size
+
+    /**
+     * Remove all pending mutations and run them to bring state up to date
+     */
+    fun update() {
+        pendingMutations.forEach { it.invoke(state) }
+        pendingMutations.clear()
+    }
+
+    /**
+     * Push a pending mutation
+     */
+    fun mutate(mutation: StateMutation) { pendingMutations.add(mutation) }
+
+    /**
+     * Get the top of the state stack
+     */
+    fun current(): LexerState = state.top()
+}
+
+/**
+ * Tokenizes JSON documents
+ */
+internal class JsonLexer(
+    private val data: ByteArray
+) : JsonStreamReader {
+    private var peeked: JsonToken? = null
+    private val state = StateManager()
+    private var idx = 0
+
+    override suspend fun nextToken(): JsonToken {
+        val next = peek()
+        peeked = null
+        state.update()
+        return next
+    }
+
+    override suspend fun peek(): JsonToken = peeked ?: doPeek().also { peeked = it }
```

Comment:
:pinching_hand:

Reply:
?

Reply:
was the closest thing I could make to `chefs-kiss`